packages/react-core/src/components/Button/Button.tsx

```diff
@@ -55,34 +59,66 @@ export const Button: React.FunctionComponent<ButtonProps & OUIAProps> = ({
   isActive = false,
   isBlock = false,
   isDisabled = false,
+  isAriaDisabled = false,
+  inoperableEvents = ['onClick', 'onKeyPress'],
   isInline = false,
   type = ButtonType.button,
   variant = ButtonVariant.primary,
   iconPosition = 'left',
   'aria-label': ariaLabel = null,
   icon = null,
   ouiaId = null,
-  tabIndex = null as number,
+  tabIndex = null,
   ...props
 }: ButtonProps & OUIAProps) => {
   const Component = component as any;
   const isButtonElement = Component === 'button';
+
+  let hasWarnBeta = false;
+
+  if (isAriaDisabled && process.env.NODE_ENV !== 'production' && !hasWarnBeta) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      'You are using a beta component feature (isAriaDisabled). These api parts are subject to change in the future.'
+    );
+    hasWarnBeta = true;
+  }
+
+  const preventedEvents = inoperableEvents.reduce(
+    (handlers, eventToPrevent) => ({
+      ...handlers,
+      [eventToPrevent]: (event: React.SyntheticEvent<HTMLButtonElement>) => {
+        event.preventDefault();
+      }
+    }),
+    {}
+  );
+
+  const getDefaultTabIdx = () => {
+    if (isDisabled) {
+      return isButtonElement ? null : -1;
+    } else if (isAriaDisabled) {
+      return null;
+    }
```

Comment:
How about just mutating `tabIndex`?
 ```
if (tabIndex !== null) {
  tabIndex = (isAriaDisabled || (isDisabled && isButtonElement) ? null : -1);
}`

Reply:
We generally try to avoid mutations unless necessary, what's the goal of the suggestion though? It looks like something you'd get from a truth table and isn't very readable. This optimization may save a few characters but makes it difficult to understand the code's purpose in the process. Also, this prevents the user from being able to set the tabindex attribute explicitly, which would break a couple of the tests attached.

Reply:
I think a single if statement with mutation is easier to follow than a ternary plus an arrow function with 2 control flows and another ternary.

I don't see any reason to "try to avoid mutations unless necessary." Setting default props is a mutation. Since we want to set a default prop based on other props, I think it's fine to do it as a mutation.

As for the logic, you're correct. The code should read
```js
if (tabIndex === null) {
 tabIndex = (isAriaDisabled || (isDisabled && isButtonElement) ? null : -1);
}
```

You could even do `tabIndex = getDefaultTabIndex(props)` in the prop destructuring and move `getDefaultTabIndex` up in scope. I think that reads nicely in the docs and is explicit.

Reply:
With this code, wouldn't all buttons that don't explicitly specify a tabIndex end up with `tabIndex={-1}`?

Reply:
Thanks for discussion around this topic.  Let's just not make these updates at the moment.  The logic works in its current form, and each person has a different style of coding.  We can discuss a coding style document for future updates if we want to have a best practices guideline.